File: plugins/maven/testFramework/src/com/intellij/maven/testFramework/MavenExecutionTestCase.java
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.maven.testFramework;

import com.intellij.openapi.application.WriteAction;
import com.intellij.openapi.projectRoots.JavaSdk;
import com.intellij.openapi.projectRoots.ProjectJdkTable;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.projectRoots.impl.SdkConfigurationUtil;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.newvfs.impl.VfsRootAccess;
import com.intellij.testFramework.IdeaTestUtil;

import java.io.File;

public abstract class MavenExecutionTestCase extends MavenMultiVersionImportingTestCase {

  private static final String JDK_NAME = "MavenExecutionTestJDK";
  private String myJdkHome;

  @Override
  public void setUp() throws Exception {
    edt(() -> {
      myJdkHome = IdeaTestUtil.requireRealJdkHome();
      VfsRootAccess.allowRootAccess(getTestRootDisposable(), myJdkHome);
      super.setUp();

      WriteAction.runAndWait(() -> {
        Sdk oldJdk = ProjectJdkTable.getInstance().findJdk(JDK_NAME);
        if (oldJdk != null) {
          ProjectJdkTable.getInstance().removeJdk(oldJdk);
        }
        VirtualFile jdkHomeDir = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(new File(myJdkHome));
        Sdk jdk = SdkConfigurationUtil.setupSdk(new Sdk[0], jdkHomeDir, JavaSdk.getInstance(), true, null, JDK_NAME);
        assertNotNull("Cannot create JDK for " + myJdkHome, jdk);
        ProjectJdkTable.getInstance().addJdk(jdk);
        ProjectRootManager projectRootManager = ProjectRootManager.getInstance(myProject);
        if (projectRootManager.getProjectSdk() == null) {
          projectRootManager.setProjectSdk(jdk);
        }
      });
    });
  }

  @Override
  public void tearDown() throws Exception {
    edt(() -> {
      if (myJdkHome == null) {
        //super.setUp() wasn't called
        return;
      }
      Sdk jdk = ProjectJdkTable.getInstance().findJdk(JDK_NAME);
      if (jdk != null) {
        WriteAction.runAndWait(() -> ProjectJdkTable.getInstance().removeJdk(jdk));
      }

      super.tearDown();
    });
  }
}

File: plugins/maven/testFramework/src/com/intellij/maven/testFramework/MavenExecutionTestCase.kt
// Copyright 2000-2026 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.maven.testFramework

import com.intellij.maven.testFramework.fixtures.MavenImportingTestFixture
import com.intellij.maven.testFramework.fixtures.mavenImportingFixture
import com.intellij.maven.testFramework.utils.MavenProjectJDKTestFixture
import com.intellij.openapi.application.WriteAction
import com.intellij.testFramework.junit5.TestApplication
import com.intellij.testFramework.runInEdtAndWait
import com.intellij.util.ThrowableRunnable
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach

@TestApplication
abstract class MavenExecutionTestCase(mavenVersion: String, modelVersion: String) {
  protected val maven: MavenImportingTestFixture by mavenImportingFixture(
    mavenVersion = mavenVersion,
    modelVersion = modelVersion,
  )

  private lateinit var jdkFixture: MavenProjectJDKTestFixture

  @BeforeEach
  fun setUpMavenExecutionJdk() {
    jdkFixture = MavenProjectJDKTestFixture(maven.project, JDK_NAME)
    runInEdtAndWait {
      WriteAction.runAndWait<RuntimeException?>(ThrowableRunnable { jdkFixture.setUp() })
    }
  }

  @AfterEach
  fun tearDownMavenExecutionJdk() {
    runInEdtAndWait {
      WriteAction.runAndWait<RuntimeException?>(ThrowableRunnable { jdkFixture.tearDown() })
    }
  }

  companion object {
    private const val JDK_NAME = "MavenExecutionTestJDK"
  }
}
